### PR TITLE
Add support for remote mdns candidates

### PR DESCRIPTION
**Pull request by mickel8**
related with membraneframework/membrane_demo#75

### CODE REVIEW

**Comment by mickel8** [comment]
depends on rosetta-home/mdns#19

File: lib/ex_libnice.ex [hunk 1]
```diff
@@ -390,18 +400,31 @@ defmodule ExLibnice do
...
+    address = Enum.at(candidate_sp, 4)
+
+    if String.ends_with?(address, ".local") do
+      Logger.debug("Sending query to resolve mDNS address #{inspect(address)}")
+      Mdns.Client.query(address)
+
+      state = %{
+        state
+        | mdns_queries: Map.put(state.mdns_queries, address, {candidate, stream_id, component_id})
+      }
```
Comment (mat-hek) [inline lib/ex_libnice.ex:423]: use `put_in`

File: lib/ex_libnice.ex [hunk 2]
```diff
@@ -515,6 +538,33 @@ defmodule ExLibnice do
     {:noreply, state}
   end
 
+  @impl true
+  def handle_info({_namespace, %Mdns.Client.Device{} = dev} = msg, state) do
+    Logger.debug("mDNS address resolved #{inspect(msg)}")
+
+    state =
+      case Map.pop(state.mdns_queries, dev.domain) do
```
Comment (mat-hek) [inline lib/ex_libnice.ex:546]:
```suggestion

    {query, state} = pop_in(state, [:mdns_queries, dev.domain])
      case query do
```

File: lib/ex_libnice.ex [hunk 3]
```diff
@@ -515,6 +538,33 @@ defmodule ExLibnice do
...
+
+          %{state | mdns_queries: mdns_queries}
+
+        {{candidate, stream_id, component_id}, mdns_queries} ->
+          candidate_parts =
+            String.split(candidate, " ", parts: 6)
+            |> List.replace_at(4, :inet.ntoa(dev.ip))
+
+          candidate = Enum.join(candidate_parts, " ")
+          do_set_remote_candidate(candidate, stream_id, component_id, state)
```
Comment (mat-hek) [inline lib/ex_libnice.ex:561]: you ignore the result, is that intended?

Comment (mickel8) [reply lib/ex_libnice.ex]: Yes, this function by default returns response for handle_call but here we are in handle_info so we ignore returned response and send `{:noreplt, state}`.

File: lib/ex_libnice.ex [hunk 4]
```diff
@@ -410,21 +413,19 @@ defmodule ExLibnice do
...
   @impl true
   def handle_call({:set_remote_candidate, candidate, stream_id, component_id}, _from, state) do
-    candidate_sp = String.split(candidate, " ")
-    address = Enum.at(candidate_sp, 4)
+    candidate_sp = String.split(candidate, " ", parts: 6)
 
-    if String.ends_with?(address, ".local") do
+    withl address_error: address when address != nil <- Enum.at(candidate_sp, 4),
+          parse_error: 6 <- length(candidate_sp),
+          not_mdns: true <- String.ends_with?(address, ".local") do
```
Comment (mat-hek) [inline lib/ex_libnice.ex:420]:
```suggestion
    withl candidate_check: 6 <- length(candidate_sp),
          do: address = Enum.at(candidate_sp, 4),
          mdns_check: true <- String.ends_with?(address, ".local") do
```

File: lib/ex_libnice.ex [hunk 5]
```diff
@@ -410,21 +413,19 @@ defmodule ExLibnice do
...
-        | mdns_queries: Map.put(state.mdns_queries, address, {candidate, stream_id, component_id})
-      }
-
+      state = Struct.put_in(state, [:mdns_queries, address], {candidate, stream_id, component_id})
       {:reply, :ok, state}
     else
-      do_set_remote_candidate(candidate, stream_id, component_id, state)
+      address_error: _ -> {:reply, {:error, :failed_to_parse_sdp_string}, state}
+      parse_error: _ -> {:reply, {:error, :failed_to_parse_sdp_string}, state}
+      not_mdns: _ -> do_set_remote_candidate(candidate, stream_id, component_id, state)
```
Comment (mat-hek) [inline lib/ex_libnice.ex:428]:
```suggestion
      candidate_check: _ -> {:reply, {:error, :failed_to_parse_sdp_string}, state}
      mdns_check: false -> do_set_remote_candidate(candidate, stream_id, component_id, state)
```

**Review by mat-hek — APPROVED** [review]

**Pull request merged**